### ISSUE TITLE
Catboost Optuna halt for no reason after few trails

### ISSUE THREAD

**Issue by RaistlinTAO** [optuna/optuna#4295]
### Expected behavior

The optuna should finish optimize() based on n_trials

### Environment

- Optuna version: Latest
- Python version: 3.10
- OS: CentOS
- (Optional) Other libraries and their versions:
- Catboost: Latest

Hardware:
 i9 9900K 
 1GB Memory

### Error messages, stack traces, or logs

```shell
[I 2023-01-02 18:50:16,041] Trial 241 finished with value: 0.0011860588536081993 and parameters: {'loss_function': 'MultiRMSE', 'n_estimators': 1493, 'learning_rate': 0.025969175404196965, 'l2_leaf_reg': 2.27279427709343, 'bootstrap_type': 'MVS', 'max_depth': 15, 'grow_policy': 'SymmetricTree', 'rsm': 0.4331822919262289, 'subsample': 0.25207733110609043, 'sampling_frequency': 'PerTree'}. Best is trial 183 with value: 0.0006216954086151799.

................

1487:     learn: 0.0069663        total: 10m 36s  remaining: 2.14s
1488:     learn: 0.0069587        total: 10m 37s  remaining: 1.71s
1489:     learn: 0.0069528        total: 10m 37s  remaining: 1.28s
1490:     learn: 0.0069435        total: 10m 37s  remaining: 856ms
1491:     learn: 0.0069377        total: 10m 38s  remaining: 428ms
1492:     learn: 0.0069318        total: 10m 38s  remaining: 0us

then the whole process finished without any error meessage.
```

### Steps to reproduce

1. Prepare Catboost objective
2. Running the code below
3. Optuna halt for no reason after finished several trails (this seems random, it may halt at 200, 20, 60 trails)
4. The log indicates 0us left in that trail

```console
1492:     learn: 0.0069318        total: 10m 38s  remaining: 0us
```

```python
def objective(trial):
    # Code for data fetching
    X_train, X_test, y_train, y_test = train_test_split()
    # Scale the data
    scaler = StandardScaler()
    X_train = scaler.fit_transform(X_train)
    X_test = scaler.transform(X_test)

    param = {
        "loss_function": trial.suggest_categorical("loss_function",
                                                   ["RMSE", "MultiRMSE", "MAE", "Quantile:alpha=0.5",
                                                    "LogLinQuantile:alpha=0.5", "Poisson", "MAPE", "Lq:q=1"]),
        "iterations": trial.suggest_int("n_estimators", 100, 2000),
        "learning_rate": trial.suggest_float("learning_rate", 0.00001, 0.2),
        "reg_lambda": trial.suggest_float("l2_leaf_reg", 0.0, 10.0),
        "bootstrap_type": trial.suggest_categorical("bootstrap_type", ["Bayesian", "Bernoulli", "MVS"]),
        "depth": trial.suggest_int('max_depth', 3, 16),
        "grow_policy": trial.suggest_categorical("grow_policy", ["SymmetricTree", "Depthwise", "Lossguide"]),
        "rsm": trial.suggest_float("rsm", 0.0, 1.0)
    }

    # Other params

    model = cb.CatBoostRegressor(**param)
    model.fit(X_train, y_train, early_stopping_rounds=100)
    y_pred = model.predict(X_test)
    return mean_squared_error(y_test, y_pred)

if __name__ == '__main__':
    study = optuna.create_study(direction='minimize')
    study.optimize(objective, n_trials=100)
    # CODE FOR RESULT PROCESSING
```

### Additional context (optional)

I think it maybe related to the memory limit so the process get force terminated, therefore a failsafe should be implemented here.

**Comment by RaistlinTAO** [comment]
> Probably the issue comes from catboost. Could you share minimal "reproducible" codes with us? The code doesn't have import statements, dataset definitions, and no random seed fixing. Without such reproducible code, it is hard to investigate the issue for maintainers.

```python
import os
import sys
import warnings

import catboost as cb
import numpy as np
import optuna
import pandas as pd
from optuna.distributions import CategoricalDistribution, IntUniformDistribution, UniformDistribution
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

__location__ = os.path.realpath(os.path.join(os.getcwd(), os.path.dirname(__file__)))

raw_data = pd.read_csv(os.path.join(__location__, '../data/all.csv'))
x = raw_data.loc[:, ["a", "b", "c",
                                 "d", "e", "f",
                                 "g", "h", "i"]]
y = raw_data.loc[:, "z"]
X_train, X_test, y_train, y_test = train_test_split(x, y, train_size=0.90, test_size=0.10)
```

Data example:
a,b,c, d,e,f, g,h,i, z
0.029, 26.14, -0.1, 0.510266224, 0.631091698, 0.630647322, 0.423781085, 0.365112859, 0.366444707,
0.142840854
